src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/CacheTagKey.cs

```diff
@@ -0,0 +1,319 @@
+            var key = GenerateKey();
+
+            // The key is typically too long to be useful, so we use a cryptographic hash
+            // as the actual key (better randomization and key distribution, so small vary
+            // values will generate dramatically different keys).
+            using (var sha = SHA256.Create())
+            {
+                var contentBytes = Encoding.UTF8.GetBytes(key);
+                var hashedBytes = sha.ComputeHash(contentBytes);
+                return Convert.ToBase64String(hashedBytes);
+            }
+        }
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            if (obj == null || !(obj is CacheTagKey))
+            {
+                return false;
+            }
+
+            return Equals((CacheTagKey)obj);
+        }
+
+        /// <inheritdoc />
+        public bool Equals(CacheTagKey other)
+        {
+            return string.Equals(other._key, _key, StringComparison.Ordinal) &&
+                other._expiresAfter == _expiresAfter &&
+                other._expiresOn == _expiresOn &&
+                other._expiresSliding == _expiresSliding &&
+                string.Equals(other._varyBy, _varyBy, StringComparison.Ordinal) &&
+                AreSame(_cookies, other._cookies) &&
+                AreSame(_headers, other._headers) &&
+                AreSame(_queries, other._queries) &&
+                AreSame(_routeValues, other._routeValues) &&
+                _varyByUser == other._varyByUser &&
+                (!_varyByUser || string.Equals(other._username, _username, StringComparison.Ordinal)
+                );
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            // The hashcode is intentionally not using the computed 
+            // stringified key in order to prevent string allocations
+            // in the common case where it's not explicitly required.
+
+            // Caching as the key is immutable and it can be called
+            // multiple times during a request.
+            if (_hashcode.HasValue)
+            {
+                return _hashcode.Value;
+            }
+
+            var hashCodeCombiner = new HashCodeCombiner();
+
+            hashCodeCombiner.Add(_key, StringComparer.Ordinal);
+            hashCodeCombiner.Add(_expiresAfter);
+            hashCodeCombiner.Add(_expiresOn);
+            hashCodeCombiner.Add(_expiresSliding);
+            hashCodeCombiner.Add(_varyBy, StringComparer.Ordinal);
+            hashCodeCombiner.Add(_username, StringComparer.Ordinal);
+
+            CombineCollectionHashCode(hashCodeCombiner, VaryByCookieName, _cookies);
+            CombineCollectionHashCode(hashCodeCombiner, VaryByHeaderName, _headers);
+            CombineCollectionHashCode(hashCodeCombiner, VaryByQueryName, _queries);
+            CombineCollectionHashCode(hashCodeCombiner, VaryByRouteName, _routeValues);
+
+            _hashcode = hashCodeCombiner;
```

Comment:
Be explicit instead of relying on converters. `_hashcode = hashCodeCombiner.CombinedHash;`